staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -18,37 +18,51 @@ package shufflesharding
 
 import (
 	"errors"
+	"fmt"
 	"math"
+	"strings"
 )
 
 const maxHashBits = 60
 
-// ValidateParameters can validate parameters for shuffle sharding
-// in a fast but approximate way, including deckSize and handSize
-// Algorithm: maxHashBits >= bits(deckSize^handSize)
-func ValidateParameters(deckSize, handSize int32) bool {
-	if handSize <= 0 || deckSize <= 0 || handSize > deckSize {
-		return false
+// ValidateParameters finds errors in the parameters for shuffle
+// sharding.  Returns a slice for which `len()` is 0 if and only if
+// there are no errors.  The entropy requirement is evaluated in a
+// fast but approximate way: bits(deckSize^handSize).
+func ValidateParameters(deckSize, handSize int) (errs []string) {
+	if handSize <= 0 {
+		errs = append(errs, "handSize is not positive")
 	}
-
-	return math.Log2(float64(deckSize))*float64(handSize) <= maxHashBits
+	if deckSize <= 0 {
+		errs = append(errs, "deckSize is not positive")
+	}
+	if len(errs) > 0 {
+		return
+	}
+	if handSize > deckSize {
+		return []string{"handSize is greater than deckSize"}
+	}
+	if math.Log2(float64(deckSize))*float64(handSize) > maxHashBits {
+		return []string{fmt.Sprintf("more than %d bits of entropy required", maxHashBits)}
+	}
+	return
 }
 
 // ShuffleAndDeal can shuffle a hash value to handSize-quantity and non-redundant
 // indices of decks, with the pick function, we can get the optimal deck index
 // Eg. From deckSize=128, handSize=8, we can get an index array [12 14 73 18 119 51 117 26],
 // then pick function will choose the optimal index from these
 // Algorithm: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#queue-assignment-proof-of-concept
-func ShuffleAndDeal(hashValue uint64, deckSize, handSize int32, pick func(int32)) {
-	remainders := make([]int32, handSize)
+func ShuffleAndDeal(hashValue uint64, deckSize, handSize int, pick func(int)) {
```

Comment:
@MikeSpreitzer @yue9944882  why do we change int32 to int here?

Reply:
I documented that in the commit comment.  See it at https://github.com/kubernetes/kubernetes/pull/81182/commits/aadd486973b0bd697eaa1568e73408b52392bafc , for example.

Reply:
I also wrote about it in another comment: https://github.com/kubernetes/kubernetes/pull/81182#discussion_r313673013

Reply:
got it, I think both are OK if we accept the type conversion in priority and fairness. cc @yue9944882

Reply:
convinced, let's make it an int